--- compiler-rt-10.0.0.src/cmake/Modules/AddCompilerRT.cmake
+++ compiler-rt/cmake/Modules/AddCompilerRT.cmake
@@ -615,6 +615,8 @@
                -DLLVM_LIBRARY_OUTPUT_INTDIR=${prefix}/lib
                -DCOMPILER_RT_LIBCXX_PATH=${COMPILER_RT_LIBCXX_PATH}
                -DCOMPILER_RT_LIBCXXABI_PATH=${COMPILER_RT_LIBCXXABI_PATH}
+               -DLIBCXX_HAS_MUSL_LIBC=ON
+               -DLIBCXX_HAS_GCC_S_LIB=OFF
                ${LIBCXX_CMAKE_ARGS}
     INSTALL_COMMAND ""
     STEP_TARGETS configure build
