diff --git a/cmake/Modules/AddCompilerRT.cmake b/cmake/Modules/AddCompilerRT.cmake
index cd4c704..dcfab0d 100644
--- a/cmake/Modules/AddCompilerRT.cmake
+++ b/cmake/Modules/AddCompilerRT.cmake
@@ -575,6 +575,7 @@ macro(add_custom_libcxx name prefix)
     DEPENDS ${name}-clobber ${LIBCXX_DEPS}
     PREFIX ${prefix}
     SOURCE_DIR ${COMPILER_RT_LIBCXX_PATH}
+    BUILD_COMMAND /usr/bin/cmake --build . -- -j1
     STAMP_DIR ${STAMP_DIR}
     BINARY_DIR ${BINARY_DIR}
     CMAKE_ARGS ${CMAKE_PASSTHROUGH_VARIABLES}
@@ -586,6 +587,8 @@ macro(add_custom_libcxx name prefix)
                -DLLVM_BINARY_DIR=${prefix}
                -DLLVM_LIBRARY_OUTPUT_INTDIR=${prefix}/lib
                -DLIBCXX_STANDALONE_BUILD=ON
+               -DLIBCXX_HAS_MUSL_LIBC=ON
+               -DLIBCXX_HAS_GCC_S_LIB=OFF
                ${LIBCXX_CMAKE_ARGS}
     INSTALL_COMMAND ""
     STEP_TARGETS configure build
