diff --git a/CMakeLists.txt b/CMakeLists.txt
index 21d0c92..3a137aa 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -446,7 +446,7 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${LIBCXX_LIBRARY_DIR})
 # LIBCXX_LIBRARIES: libraries libc++ is linked to.
 set(LIBCXX_COMPILE_FLAGS "")
 set(LIBCXX_LINK_FLAGS "")
-set(LIBCXX_LIBRARIES "")
+set(LIBCXX_LIBRARIES "-lgcc")
 
 # Include macros for adding and removing libc++ flags.
 include(HandleLibcxxFlags)
