diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5df3f25..6b62484 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -449,7 +449,7 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${LIBCXX_LIBRARY_DIR})
 #                             These libraries are exposed in the linker script.
 set(LIBCXX_COMPILE_FLAGS "")
 set(LIBCXX_LINK_FLAGS "")
-set(LIBCXX_LIBRARIES "")
+set(LIBCXX_LIBRARIES "-lgcc")
 set(LIBCXX_INTERFACE_LIBRARIES "")
 
 # Include macros for adding and removing libc++ flags.
