From 4f9a8ae81b7c3810b1ec8e31fc11cf72c0f82dee Mon Sep 17 00:00:00 2001
From: Dmitry Golovin <dima@golovin.in>
Date: Wed, 19 Jul 2017 13:35:25 +0300
Subject: [PATCH] prefer Oz optimization if the compiler supports it

---
 configure | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure b/configure
index 539c9fcb..1212af91 100755
--- a/configure
+++ b/configure
@@ -425,7 +425,7 @@ xno|x) printf "disabled\n" ; optimize=no ;;
 *) printf "custom\n" ;;
 esac
 
-test "$optimize" = no || tryflag CFLAGS_AUTO -Os || tryflag CFLAGS_AUTO -O2
+test "$optimize" = no || tryflag CFLAGS_AUTO -Oz || tryflag CFLAGS_AUTO -Os || tryflag CFLAGS_AUTO -O2
 test "$optimize" = yes && optimize="internal,malloc,string"
 
 if fnmatch 'no|size' "$optimize" ; then :
-- 
2.13.1

